src/display.rs

```diff
@@ -310,13 +351,39 @@ where
     <S as ToOwned>::Owned: fmt::Debug,
     &'a S: AsRef<[u8]>,
 {
+    // write the part within the styling prefix and suffix
+    fn write_inner<W: AnyWrite<Wstr = S> + ?Sized>(&self, w: &mut W) -> Result<(), W::Error> {
+        // \x01 .. \x02 are used to wrap zero-width portions of output
+        match &self.oscontrol {
+            Some(OSControl::Link { url: u }) => {
+                write!(w, "\x01\x1B]8;;")?;
+                w.write_str(u.as_ref())?;
+                write!(w, "\x1B\x5C\x02")?;
+                w.write_str(self.string.as_ref())?;
+                write!(w, "\x01\x1B]8;;\x1B\x5C\x02")
+            }
+            Some(OSControl::Title) => {
+                write!(w, "\x01\x1B]2;")?;
+                w.write_str(self.string.as_ref())?;
+                write!(w, "\x1B\x5C\x02")
```

Comment:
I'm not sure what these ending chars are (\x5C\x02). I thought Title, Icon, Title-with-Icon, and Hyperlink were all supposed to end with BEL (0x07) or ST (0x9C)? Am I misunderstanding?

Also, Title-with-Icon is missing I think
https://tldp.org/HOWTO/Xterm-Title-3.html

Hyperlink info https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda

Reply:
My primary reference has been the [control sequences from xterm](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Operating-System-Commands) which is also consistent with the references you linked.

> @fdncred
> I'm not sure what these ending chars are (\x5C\x02)

The first byte is "\". The second is second part of a pair that wraps zero-width sequences. So when the terminal sees \x01 ... \x02 it knows the ... portion shouldn't be made visible. Incidentally these are called:
0x01 SOH Start of Heading
0x02 STX Start of Text

I've been testing these sequences in prompt strings and without those zero-width markings it messes up libraries like readline. If you look at the newly emitted sequences in this PR you should see that most start with 0x01 because, with the exception of a portion of the hyperlink, most OSC sequences in this PR are not meant to be displayed in the terminal.

> @fdncred
> I thought Title, Icon, Title-with-Icon, and Hyperlink were all supposed to end with BEL (0x07) or ST (0x9C)? Am I misunderstanding?

You are correct. However 0x9C (notice it's > 0x7F) is part of the "C1" section and from what I've read everything in the C1 section tends to conflict with non-ASCII character encodings (UTF-8 and UTF-16) since they use values above 0x7F. So I've chosen to stick to the "C0" section to maximize portability and/or avoid needing to detect the capabilities and mode of the current terminal.

As for the question of using BEL vs ST, the xterm docs on control sequences state that it supports both for OSC sequences. The PR is using ST which is also used in the hyperlink example found in the gist you linked.

> @fdncred 
> Also, Title-with-Icon is missing I think

Yes, and I omitted it for a couple reasons.

Choosing the specific ones that only do one or the other -- set the icon or title but not both -- is probably more portable. Also it's pretty unclear what it means to set the title and icon name to the same thing at the same time so I figure that ambiguity might be good to avoid engaging with when defining an API.

If users of the API want to set both that can call them separately.

Also, if it proves to be needed in the future the API can still be extended to support that.

[Regarding portability, this TLDP page says:](https://tldp.org/HOWTO/Xterm-Title-6.html)
> "The Irix terminals also support the xterm escapes to individually set window title and icon title, but not the escape to set both."

--
The xterm reference I gave is a bit long so I thought it might help to quote useful parts here.

In the OSC section it says:
> XTerm accepts either BEL  or ST  for terminating OSC

It also helpfully describes the trouble with using C1 codes in modern systems.
> ECMA-48 (aka "ISO 6429") documents C1 (8-bit) and C0 (7-bit) codes.
> Those are respectively codes 128 to 159 and 0 to 31.
...
> Controls (including the escape code 27) are processed once:
> 
> o   This means that a C1 control can be mistaken for badly-formed UTF-8
>      when the terminal runs in UTF-8 mode because C1 controls are valid
>      continuation bytes of a UTF-8 encoded (multibyte) value.

Reply:
Thanks for the thorough explanation, it helps.

I'm all for better portability and generally agree with your choices. However, I'm concerned about 0x01 and 0x02 bits. What precedent to you have for using this? I didn't see them mentioned in the xterm doc you referenced. I've gone through that thing many times myself but may have just missed it.

I have noticed that some terminals respond better to BEL than ST or vice versa. I wonder if SOH/STX would fix those situations. Not asking you to do any testing or anything but your comments made me think about it.

Reply:
> Thanks for the thorough explanation, it helps.

Great! I tend to look at this as part of documenting the changes so I err on the side of being thorough. It's especially important because some of this is rather obscure and it's taken me a while to dig it up, check it in references for multiple terminal emulators.

> I'm all for better portability and generally agree with your choices. However, I'm concerned about 0x01 and 0x02 bits. What precedent to you have for using this? I didn't see them mentioned in the xterm doc you referenced. I've gone through that thing many times myself but may have just missed it.

Ah, good point. That comes from documentation surrounding prompt generation. It's used indirectly in starship in the form of `\[` and `\]` (Note that the backslashes appear in the strings -- they're not escaping the square brackets) when building prompts for bash. Bash then expands strings, when found in $PS1, into `0x01` and `0x02` respectively ( \#define for  RL_PROMPT_START_IGNORE and RL_PROMPT_END_IGNORE). According to the starship code for zsh and tcsh those shells similarly translate the strings `%{` and `}%` when expanding prompts. That being said, I don't see any way for the terminal emulator to know these sequences are part of shell prompts so I think it treats anything wrapped in 0x01 ... 0x02 as zero-width.

In some documentation they're referred to as `^A` and `^B`

https://en.wikipedia.org/wiki/C0_and_C1_control_codes#SOH

In the bash source code, lib/readline/ there's this (prefixed with line numbers):
```
./readline.h
885:#define RL_PROMPT_START_IGNORE	'\001'

./display.c
341: \001 (^A) start non-visible characters
342:           \002 (^B) end non-visible characters
343:     all characters except \001 and \002 (following a \001) are copied to
344:     the returned string; all characters except those between \001 and
345:     \002 are assumed to be `visible'.
```

In the context of .inputrc (which at least readline uses see: https://medium.com/everything-full-stack/terminal-escape-codes-are-awesome-heres-why-c8eb938b1a1c ):
```
# \1 — Start zero-width character (\001)
# \2 — End zero-width character (\002)
# \e — By itself simply means “an escape character”
# \e[ — Signal CSI (Control Sequence Introducer)
# \e] — Signal OSC (Operating System Command)
```

Also another helpful resource for wezterm (which I've seen references to but never used): https://wezfurlong.org/wezterm/escape-sequences.html#c0-control-codes

suggests these are "ignored".

It's a little unsettling that these aren't *documented* in the xterm references but I have *tested* them under xterm and, as best I can tell, they work exactly as expected. I even looked to see if they're documented in the vt100 references I could find and no such luck.

> I have noticed that some terminals respond better to BEL than ST or vice versa. I wonder if SOH/STX would fix those situations. Not asking you to do any testing or anything but your comments made me think about it.

Hmm, interesting. I've just been manually testing in gnome-terminal, xterm, eterm, rxvt, and alacritty. I only have Linux systems so I can't test iTerm2 or windows terminal software. So far they all seem to support ST but it's also quite likely I've missed some combinations (os + term) and I'm curious to get a sense of what those are if you can recall specific ones I can try.

One interesting related part of  the wezterm reference says:
```
OSC sequences are canonically delimited by the ST (String Terminator) sequence, but WezTerm also accepts delimiting them with the BEL control.
```

Reply:
TIL - All this stuff. Thanks for the education! :)

You last few points hit on where I was going, which is being cross-platform friendly. We forked this crate to update it a little but to also allow it to work in nushell well. Nushell runs on MacOS, Linux, and Windows (amongst others). So, we try our best to support these major 3 platforms everywhere.

Here's a list of terminals I personally try to test with/use off and on, although I'm suer that forgetting a few.

WezTerm - MacOS (still kind of broken in Windows)
Alacritty - MacOS/Linux
Windows Terminal - Windows
gnome terminal - Linux
Terminal.app - MacOS
Tabby - MacOS/Windows
Kitty - MacOS/Linux
iTerm2 - MacOS
Visual Studio Code - MacOS/Linux/Windows

So, we really need to test a few on the platforms that you haven't tested. Any tips on how to do that?

Reply:
> So, we really need to test a few on the platforms that you haven't tested. Any tips on how to do that?

Well, the way I've been testing is I use a branch of Starship as my command line prompt generator. It tests out the hyperlink bits every time I enter a new command on any of my open terminals.

As for `::title()` and `::cwd()`: I've been testing those sequences more manually by echo-ing the output strings but I'm hoping to add bits to Starship for those as well.

`::icon()` is trickier to test since I'm not aware of anything that pays attention to the "icon title" any more. I think this was used in Ye Olden Days when X11 windows could be minimized into an icon above some text that would be used to restore it. In modern Linux desktops (e.g. Wayland-based rather than X.org) I'm not sure there are corresponding elements to set much less if they ever get displayed somewhere.

One difficulty is a lot of command line prompts set the window title so to test that manually or via cargo you've got to disable that (part of) your prompt generator. So while I've been tempted to write an example testcase much like the table of colors I haven't (yet) thought of a way to temporarily disable the prompt resetting it before you can see the result.

Something like PhantomJS but for terminals sure would be a Nice-to-Have.

Reply:
Do we have tests to cover this that run in the CI? If not, what can we do to test these features in CI? We probably also need to add macos-latest in the CI as well.

Reply:
OK, I hear you on the CI end of things.

For now I'm trying to make it easier to manually test.

First: I've written some examples much like the `cargo run --example 256_colors` in the current CI runs. The `::title()`, `::icon()`, and `::cwd()` examples sleep for 5 seconds so if you run them you have a moment to look at, reveal , or use those respectively in the terminal emulators that support them.

It's still a manual process, though, and the user needs to know what their shell does with that information. For example, on my machine (Linux, GNOME Shell, gnome-terminal), I can only see the icon title if I enter "overview" mode where I can see the running application icons and icon titles for the apps visible on that screen. However my terminals don't seem to do anything with `::cwd()` yet I've heard that Mac OS Terminal.app uses it when opening a new terminal window from an existing terminal.

So it's still very much a manual test and that's why I put these in the examples.

I'm also looking into optionally running combinations of terminal, shell, and OSC output when users run `cargo test osc`.
For example, on Linux, it could cross gnome-terminal, xterm, rxvt, Eterm, and alacritty with bash, zsh, and tcsh, emit the control sequence for each OSC function, and the tester would see the "output". Since it's starting new terminal emulator windows for each they should appear showing the correct titles (window and icon), working directory, and hyperlinks.

But I still don't see how we could get that into CI hence those tests wouldn't be run with a simple `cargo test` nor `cargo test --all-features`

Before I think about CI integration some more I'm going to get these tests finished and pushed as additional commits so you folks can see them.

Reply:
> The ::title(), ::icon(), and ::cwd() examples sleep for 5 seconds so if you run them you have a moment to look at, reveal , or use those respectively in the terminal emulators that support them.

I think that's fine for one-off testing but not for CI. We just need to ensure that the right escape sequence is being written out. As you know, terminals decide for themselves how to interpret any ansi escape we create.

>  However my terminals don't seem to do anything with ::cwd() ...

When we were working on this in nushell, what I remember about it was that I had to change a directory, for example `cd /tmp`. Then, on terminals that support tabs, I start a new tab, and it should start in `/tmp`. Like you're saying, not all terminals support that escape sequence.

So, reiterating, we can't be responsible for any terminal's response, but we are responsible to ensure that we're as compatible as possible and that we're passing whatever the compatible sequence is accurately. I feel like I'm preaching to the choir because I'm sure you know all this. 

> I'm also looking into optionally running combinations ...

For manual tests, I think it's a great idea to have some way of testing on a variety of operating systems and terminals. I don't really see that working in CI though, like you're saying. 

We could have some test code, that takes a file as a parameter, and in that file is a full path to the terminal(s) we want to test. Maybe even as an example or some type of test that doesn't run in CI but we can invoke manually. For these type of things, we probably need some manually invoked functionality that is different from CI where we just parse and check for the right escape codes.

Reply:
> We could have some test code, that takes a file as a parameter, ...

In the shower today, I realized that this suggestion was crazy-talk.

> For manual tests, I think it's a great idea to have some way of testing on a variety of operating systems and terminals. 

I still think it would be good to have an example that we can run with `cargo run --example manual_test --features=whatev` that we can manually run at will in a variety of terminals to see if the title changes and whatnot.

> So it looks like CI is limited ...

If nothing else is possible for CI, then that's fine. The only reason I'm very interested in tests is because many people depend on this crate. I'd like to get to 48.1 or 49.0 sooner rather than later to avoid the breaking change in 48.

Reply:
@fdncred 
> > For manual tests, I think it's a great idea to have some way of testing on a variety of operating systems and terminals.
> 
> I still think it would be good to have an example that we can run with `cargo run --example manual_test --features=whatev` that we can manually run at will in a variety of terminals to see if the title changes and whatnot.

Yup, that's what the examples I created do except without the `--features=whatev` part:
```
cargo run --example hyperlink
cargo run --example title
...
```

Take a look at the last commit in the PR as it stands now. Just like the `cargo run --example 256_colors` CI step it adds the `OSC` examples to the CI runs and you can see the output in the most recent CI run:
```
Run cargo run --example hyperlink
   Compiling nu-ansi-term v0.48.0 (D:\a\nu-ansi-term\nu-ansi-term)
    Finished dev [unoptimized + debuginfo] target(s) in 0.26s
     Running `target\debug\examples\hyperlink.exe`

```

Notice how the text `Link to example.com` is missing. I think that just shows the CI output processing code doesn't know what to make of the hyperlink output but it does notice the Blue + underline styling.

> 
> > So it looks like CI is limited ...
> 
> If nothing else is possible for CI, then that's fine. The only reason I'm very interested in tests is because many people depend on this crate. I'd like to get to 48.1 or 49.0 sooner rather than later to avoid the breaking change in 48.

Makes sense. There are so many people using various seemingly-obscure open source projects that don't show up in various metrics that it's unsafe to assume the few people use it. Sometimes things being quiet just means folks are happy with it and that's a good "problem" to have.

It's probably more much more involved and speculative but perhaps there are some container CI + screenshot output workflows out there I could look into.

It's almost like we want a headless terminal (xtermjs? DomTerm?) somewhat like what phantomjs used to provide for web browser code.

I'll keep looking at those two -- running in container + screenshot or browser-based terminals but it would be really nice to get some of this merged soon. For instance, that Style test commit could be cherry-picked even without the rest (though I suppose it would show CI failing until the `OSC` fixes go in. Still, it better reflects the reality of the tree as it stands now..).

Alternately: Would  it make sense to put the `OSC` API behind a feature flag? I'm not usually a huge fan of APIs changing based on a feature flags because those can proliferate to the point of being their own special config / dll / registry hell. But it is fairly distinct and a secondary feature compared to the Styling and Coloring nu-ansi-term focuses on so I'm ambivalent on this idea.

Reply:
> Yup, that's what the examples I created do ...

I'll try to look at them tonight sometime.

> It's almost like we want a headless terminal (xtermjs? DomTerm?) somewhat like what phantomjs used to provide for web browser code.

Have you looked at Reilly's [escape artist](https://github.com/rgwood/escape-artist)? You might something interesting in there,specifically around vte.

> Alternately: Would it make sense to put the OSC API behind a feature flag?

I'm not sure there's a lot of value there because there are so few commands that would be behind that feature. If we can do what you're doing here, without breaking everyone else, I think we'll be fine but I'm interested what others think (if they're watching this thread).

Reply:
OK, so I looked at xtermjs, escape artist, and libvte. The latter two would require taking screenshots and couldn't do a simple pass/fail in CI. Taking the screenshots is still not something I found a way to do nicely.

xtermjs has a headless mode and API. It can take input and use that to set its window and icon title fields, for example, which we could then check to see if they match the input we gave it.

However, xtermjs requires a slew of new code and new CI bits which will each need their own testing to become reliable. On the plus side though, whenever the CI stuff breaks fewer people will notice.

Then I found vtparse (https://github.com/haberman/vtparse). It's an incredibly tiny table-driven parser for virtual terminal input that uses Ruby to generate C code from their state machine description. It builds a library (.a) very easily and has only two function entry points -- initialization and consuming a byte of input. In turn it invokes a supplied callback indicating what's happening (e.g. OSC sequence start, OSC sequence end). It parses CSI and OSC sequences so it'd be useful for testing coloring, styling, and OSC sequences. So I've been working on a rust wrapper for it.